crates/circuit/src/dag_circuit.rs

```diff
@@ -2216,6 +2233,182 @@ impl DAGCircuit {
+    ///     operation is present, even if they are individually equal.
+    fn structurally_equal(&self, other: &DAGCircuit) -> PyResult<bool> {
+        if self.qubits != other.qubits {
+            return Ok(false);
+        }
+        if self.clbits != other.clbits {
+            return Ok(false);
+        }
+        if self.vars != other.vars {
+            return Ok(false);
+        }
+        if self.stretches != other.stretches {
+            return Ok(false);
+        }
+        if self.qregs != other.qregs {
+            return Ok(false);
+        }
+        if self.cregs != other.cregs {
+            return Ok(false);
+        }
+        // Checking the identfiier information should handle all different methods of declaration
+        // for all identfiiers, whether var or stretch.  We do a manual iteration-based check here
+        // because `IndexMap`'s default is order-insensitive, but we actually do care about order
+        // for the structural check.
+        if self.identifier_info.len() != other.identifier_info.len() {
+            return Ok(false);
+        }
+        if self
+            .identifier_info
+            .iter()
+            .zip(other.identifier_info.iter())
+            .any(|(a, b)| a != b)
+        {
+            return Ok(false);
+        }
+        // This is a stricter check than `Param::eq`, since we don't allow equality between explicit
+        // floats and Python-object representations of the same float.
+        let param_eq = |left: &Param, right: &Param| -> PyResult<bool> {
+            match (left, right) {
+                (Param::Float(a), Param::Float(b)) => Ok(a.total_cmp(b) == Ordering::Equal),
+                (Param::ParameterExpression(a), Param::ParameterExpression(b)) => Ok(a.eq(b)),
+                (Param::Obj(a), Param::Obj(b)) => Python::with_gil(|py| a.bind(py).eq(b.bind(py))),
+                _ => Ok(false),
+            }
+        };
+        if !param_eq(&self.global_phase, &other.global_phase)? {
+            return Ok(false);
+        }
+        // This is stricter than `PackedInstruction::py_op_eq` because it doesn't allow equality
+        // between `StandardGate` and Python versions of that.  Additionally, it short-circuits out
+        // of control-flow, rather than recursing through while we currently don't have a clean
+        // representation of DAG-like structured control flow.
+        let inst_eq = |from_self: &PackedInstruction,
+                       from_other: &PackedInstruction|
+         -> PyResult<bool> {
+            if self.qargs_interner.get(from_self.qubits)
+                != other.qargs_interner.get(from_other.qubits)
+            {
+                return Ok(false);
+            }
+            if self.cargs_interner.get(from_self.clbits)
+                != other.cargs_interner.get(from_other.clbits)
+            {
+                return Ok(false);
+            }
```

Comment:
Not that performance matters, but would it be faster to just build a mapping vec from other to self outside the main loop and then just check that `self.qarg == interner_map[other.qarg]` ?

Reply:
It's plausible, sure - we could follow up.